testnet/stacks-node/src/run_loop/neon.rs

```diff
@@ -394,156 +472,229 @@ impl RunLoop {
+            .expect("Unable to get burnchain tip");
+
+        // Boot up the p2p network and relayer, and figure out how many sortitions we have so far
+        // (it could be non-zero if the node is resuming from chainstate)
+        let mut node = StacksNode::spawn(
+            self,
+            Some(burnchain_tip.clone()),
+            coordinator_senders.clone(),
+            attachments_rx,
+        );
+        let sortdb = burnchain.sortdb_mut();
+        let mut sortition_db_height = RunLoop::get_sortition_db_height(&sortdb, &burnchain_config);
 
-        let mut burnchain_height = block_height;
+        // Start the runloop
+        debug!("Begin run loop");
+        self.start_prometheus();
+        self.counters.bump_blocks_processed();
+
+        let mut burnchain_height = sortition_db_height;
         let mut num_sortitions_in_last_cycle = 1;
-        let mut learned_burnchain_height = false;
 
         // prepare to fetch the first reward cycle!
-        target_burnchain_block_height = burnchain_height + pox_constants.reward_cycle_length as u64;
+        target_burnchain_block_height = burnchain_config.reward_cycle_to_block_height(
+            burnchain_config
+                .block_height_to_reward_cycle(burnchain_height)
+                .expect("BUG: block height is not in a reward cycle")
+                + 1,
+        );
 
         debug!(
             "Begin main runloop starting a burnchain block {}",
-            block_height
+            sortition_db_height
         );
 
-        let mut last_block_height = 0;
+        let mut last_tenure_sortition_height = 0;
         loop {
-            // Orchestrating graceful termination
-            if !should_keep_running.load(Ordering::SeqCst) {
+            if !self.should_keep_running.load(Ordering::SeqCst) {
                 // The p2p thread relies on the same atomic_bool, it will
                 // discontinue its execution after completing its ongoing runloop epoch.
                 info!("Terminating p2p process");
                 info!("Terminating relayer");
                 info!("Terminating chains-coordinator");
-                coordinator_senders.stop_chains_coordinator();
 
+                coordinator_senders.stop_chains_coordinator();
                 coordinator_thread_handle.join().unwrap();
-                node.relayer_thread_handle.join().unwrap();
-                node.p2p_thread_handle.join().unwrap();
+                node.join();
 
                 info!("Exiting stacks-node");
                 break;
             }
 
+            let remote_chain_height = burnchain.get_headers_height();
+
             // wait for the p2p state-machine to do at least one pass
-            debug!("Wait until we reach steady-state before processing more burnchain blocks...");
+            debug!("Wait until we reach steady-state before processing more burnchain blocks (chain height is {}, we are at {})...", remote_chain_height, burnchain_height);
```

Comment:
This log could be more descriptive about what's happening here. What does "reach steady-state" mean?

Would this describe what's happening at this step:

```
debug!("Waiting for Stacks block downloads to reach steady-state before processing more burnchain blocks..."; "known_burn_height" => remote_chain_height, "our_burn_height" => burnchain_height);
```

But isn't this step waiting on more _Stacks_ blocks to be processed, rather than bitcoin blocks? If the log line was providing a status update, isn't the current stacks block more important than the current burnchain blocks?

Reply:
No, it's waiting for Stacks blocks.  The PoX sync watchdog observes the chainstate database and p2p network behavior and deduces when the node has downloaded all of the Stacks blocks it is going to download for the foreseeable future.  Namely, if there are some sortitions with Stacks blocks that are simply missing (and these *do exist*), the PoX sync watchdog needs to deduce that they are actually absent and never going to realistically arrive, so the node can go ahead and get the next reward cycle processed (with or without a potentially-missing anchor block).  The comment refers to the call to `pox_sync_wait()` below.

Reply:
(Btw, this whole strategy for determining when to advance from one reward cycle to another is going away once #1805 lands, since then we'll be able to infer whether or not to wait on a reward cycle's blocks based on whether or not we have the next reward cycle's anchor block on the canonical anchor block history.)

Reply:
I'll update the comment to this:

```rust
debug!("Wait until Stacks block downloads reach a quiescent state before processing more burnchain blocks"; "remote_chain_height" => remote_chain_height, "local_chain_height" => burnchain_height);
```